Filled circle and Catenary primitives

HenryWConklin opened the pull request:
For #60

HenryWConklin commented:
![Screenshot from 2022-11-10 07-30-54](https://user-images.githubusercontent.com/3664988/201137513-262d9ad7-a1f9-4366-a7a5-3b2bb50aa4f1.png)
![Screenshot from 2022-11-10 07-35-02](https://user-images.githubusercontent.com/3664988/201138083-635c4e28-b652-441d-af15-259da5c3cf18.png)

setzer22 commented on blackjack_engine/src/mesh/halfedge/primitives.rs:
```diff
@@ -337,6 +336,53 @@ impl Cylinder {
+                Vec3::new(xz.x, y, xz.y)
+            }
+        };
+        Line::build(position, segments)
```
One thing I'm thinking... :thinking: Would it be easy to analytically compute normal and tangent vectors for each point? Because if so, we could add one extra parameter to add those channels and nodes like `Extrude along curve` would work out of the box with catenary curves, which would be pretty cool.

HenryWConklin replied:
Yeah, that's pretty easy to compute. Would that go in `Line::build`? If there's an example somewhere of filling in those channels I can add that tonight/tomorrow.

setzer22 replied:
There's an example in `resample_curve` (`edit_ops.rs:1846`), but that one's probably a bit convoluted. The whole idea with mesh channels is to allow attaching arbitrary data to vertices, halfedges and faces. You would need something like:
```rust
// Create the channel, and fetch by 
let tangent_ch_id = mesh.channels.ensure_channel::<VertexId, Vec3>("tangent");
let mut tangent_ch = mesh.channels.write_channel(tangent_ch_id).unwrap();

// Then, you can use it like a map:
let v: VertexId = todo!("obtain it somehow");
tangent_ch[v] = Vec3::new(1.0, 2.0, 3.0);
```
Channels are like a sort of "loose" map in that you can always fetch from it even if you've never inserted (and you get `Default::default()` back) and you can always set a key as long as it's a valid vertex for the mesh. They're based on `slotmap::SecondaryMap`.

As you say, a good place for this code would be the `Line::build` function, you can create these channels right below the call to `write_positions` (FYI, vertex positions are just one of the channels that are allocated by default in a mesh, there's nothinc special about them other than the `read_positions` / `write_positions` convenience methods). The function could take two (perhaps optional?) functions for tangent and normal values. Feel free to refactor things a bit to make it fit, as I have the impression doing that might make the code a bit awkward :thinking:

I'm not sure what we should do with other line constructors that don't have a well-defined notion of normals (like the regular `Line` node, or `Line from points`). In those cases, we can either compute *some* normals by picking an arbitrary direction like one of the 3 main axis, or else we could make the tangent / normal channels optional and not compute them for those constructors. I'm not sure how useful the default would be for users.

setzer22 reviewed (approved):
Amazing! :smile: See my comment below for a potential improvement

HenryWConklin commented:
![Screenshot from 2022-11-11 08-47-23](https://user-images.githubusercontent.com/3664988/201390544-9078fb5c-732b-43be-a617-2ebf6b44c185.png)

HenryWConklin commented:
Had some git issues, but I think this PR is good to go now

setzer22 reviewed (commented):
Looks good :+1:  Ready to merge!

setzer22 commented:
Hmm :thinking:  I just tested this locally and while it works well, there's an annoying glitch when extruding:
![image](https://user-images.githubusercontent.com/7241990/201471548-2a260342-e464-4884-be95-a8832aef5c8d.png)

The normals look right (not shifted in the Z direction), so this could be a bug in "Extrude Along Curve". Need to investigate further, but unless you have an idea of what's going on there, I think we can merge this PR and convert this into an issue so we can tackle it later. The catenary node is super useful and I don't want to drag this more than necessary :smile:

> Also, I'll comment that there's a mismatch in the defaults for circle and extrude along curve. The circle needs to be rotated first and the faces flipped on the extrude node to get what you expect. May be a larger issue so I didn't want to mess with it too much.

Yes, I'm aware of this. I wonder what the most user-friendly solution would be. The problem seems to be the "Extrude Along Curve" node assuming the cross section is looking in the X direction, while most primitives like `Circle` or `Quad` are constructed looking up. Maybe I should just add an extra rotation matrix there to correct for this :thinking:.

setzer22 commented:
After further exploration, I think there's something fishy with those normals :thinking:. See how there's this unwanted rotation at the start that moves forward as I extend the curve?

https://user-images.githubusercontent.com/7241990/201473309-58d51fc2-41c1-44fe-888e-baaec3d5dacb.mp4

This is the code that extrude along curve uses to rotate the cross section at `edit_ops:1572` (that `dbg!` was added by me during debugging):
```rust
            let normal = normal_ch[v];
            let tangent = tangent_ch[v];
            dbg!(normal.dot(tangent));
            let cotangent = normal.cross(tangent);
            let (_, rotate, _) = glam::Affine3A::from_cols(
                cotangent.into(),
                normal.into(),
                tangent.into(),
                glam::Vec3A::ZERO,
            )
```
When using a regular `Line`, the dot is zero as expected (normal and tangent should be orthogonal). But with the catenary, I'm getting different non-zero values depending on the endpoints. Looks like the computation for the normals might be a bit off?

Now that we have gizmos, I'm going to work on some API that lets us do some debug drawing to better see what's going on, I've been meaning to do that for a while :thinking:.

HenryWConklin commented:
Fixed with some simpler code that I should have tried first. Also added a fix for when the endpoints line up vertically and the math stops working.

https://user-images.githubusercontent.com/3664988/201475860-9b59ed49-b927-4429-912c-38fafe5ceb6e.mp4

![Screenshot from 2022-11-12 05-12-55](https://user-images.githubusercontent.com/3664988/201475877-b4334abb-90f4-44c5-8d01-b194c58a93ce.png)



Let me know if you spot anything else weird.

setzer22 commented:
Amazing! Thanks for the hard work :smile:
